api/src/mailbox.rs

```diff
@@ -432,6 +528,16 @@ pub struct InvokeDpeReq {
 
 impl InvokeDpeReq {
     pub const DATA_MAX_SIZE: usize = 512;
+
+    pub fn as_bytes_partial(&self) -> &[u8] {
```

Comment:
You could eliminate a lot of code duplication with something like this:

```rs

#[repr(C)]
#[derive(Debug, AsBytes, Default, FromBytes, PartialEq, Eq)]
pub struct MailboxRespHeaderVarSize {
    pub hdr: MailboxRespHeader,
    pub data_len: u32,
}
trait ResponseVarSize: AsBytes + FromBytes {
    fn data(&self) -> CaliptraResult<&[u8]> {
        // Will panic if sizeof<Self>() is smaller than MailboxRespHeaderVarSize
        // or Self doesn't have compatible alignment with
        // MailboxRespHeaderVarSize (should be impossible if MailboxRespHeaderVarSize is the first field)
        let (hdr, data) =
            LayoutVerified::<_, MailboxRespHeaderVarSize>::new_from_prefix(self.as_bytes())
                .unwrap();
        data.get(..hdr.data_len as usize)
            .ok_or(CaliptraError::MAILBOX_API_RESPONSE_DATA_LEN_TOO_LARGE)
    }
    fn partial_len(&self) -> usize {
        let (hdr, _) =
            LayoutVerified::<_, MailboxRespHeaderVarSize>::new_from_prefix(self.as_bytes())
                .unwrap();
        size_of::<MailboxRespHeaderVarSize>() + hdr.data_len as usize
    }
    fn as_bytes_partial(&self) -> CaliptraResult<&[u8]> {
        self.as_bytes()
            .get(..self.partial_len())
            .ok_or(CaliptraError::MAILBOX_API_RESPONSE_DATA_LEN_TOO_LARGE)
    }
    fn as_bytes_partial_mut(&mut self) -> CaliptraResult<&mut [u8]> {
        let partial_len = self.partial_len();
        self.as_bytes_mut()
            .get_mut(..partial_len)
            .ok_or(CaliptraError::MAILBOX_API_RESPONSE_DATA_LEN_TOO_LARGE)
    }
}
impl<T: ResponseVarSize> Response for T {
    const MIN_SIZE: usize = size_of::<MailboxRespHeaderVarSize>() + size_of::<u32>();
}


#[repr(C)]
#[derive(Debug, AsBytes, FromBytes, PartialEq, Eq)]
pub struct InvokeDpeResp {
    pub hdr: MailboxRespHeaderVarSize,
    pub data: [u8; InvokeDpeResp::DATA_MAX_SIZE], // variable length
}

impl InvokeDpeResp {
    pub const DATA_MAX_SIZE: usize = 2200;
}
impl ResponseVarSize for InvokeDpeResp {}
```

Reply:
Thanks for writing this out. Also added a RequestVarSize trait.